File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.cpp
class Solution {
public:
  vector<TreeNode*>ret;
  unordered_map<string, int>cnt;
  string f(TreeNode* node){
    if(node == nullptr)
      return "";
    string l_string = "(" + f(node->left) + ")";
    string r_string = "(" + f(node->right) + ")";
    string curr_string =  l_string + "(" + to_string(node->val) + ")" + r_string;
    cnt[curr_string]++;
    if(cnt[curr_string] == 2)
      ret.push_back(node);
    return curr_string;
  }
  vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
    f(root);
    return ret;
  }
};

File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
    ret = []
    cnt = defaultdict(int)
    def f(node):
      if(node):
        s = f"({f(node.left)})({f(node.right)})({node.val})"
        cnt[s]+=1
        if(cnt[s] == 2):
          ret.append(node)
        return s
      return ""
    f(root)
    return ret